File: Array/LeetCode978.cpp
#include<vector>

int maxTurbulenceSize(std::vector<int>& arr)
{
    int n = arr.size();
    int lastComparison {0};
    int Endptr {0};
    int turbLen {1};
    int maxLen {1};
    int comparison;

    while (Endptr < n)
    {
        if (arr[Endptr] > arr[Endptr - 1])
        {
            comparison = 1;
        }
        else if (arr[Endptr] < arr[Endptr - 1])
        {
            comparison = -1;
        }
        else
        {
            comparison = 0;
        }
        if (comparison == -lastComparison)
        {
            turbLen++;
        }
        else if (comparison == 0)
        {
            turbLen = 1;
        }
        else
        {
            turbLen = 2;
        }
        lastComparison = comparison;
        maxLen = std::max(maxLen, turbLen);
        Endptr++;
    }
    return maxLen;
}

File: Array/GoLangDsa/SlidingWindow/Leetcode978.go
package main

func compare (a,b int) int {

	if a < b {

		return -1
	}else if a > b {

		return 1
	}

	return 0
}

func max (a,b int) int {

	if a > b {

		return a
	}

	return b
}

func maxTurbulenceSize(arr []int) int{

	length := 1
	maxLen := 1
	prevCmp := 0

	for i := 1;i < len(arr);i++ {

		currCmp := compare(arr[i - 1],arr[i])

		if currCmp == 0 {

			length = 1
		}else if prevCmp * currCmp == -1 {

			length++
		}else {

			length = 2
		}

		maxLen = max(maxLen,length)
		prevCmp = currCmp
	}

	return maxLen
}
